consensus/milestone_tracker/milestone_tracker.c

```diff
@@ -25,88 +25,91 @@
 static retcode_t validate_milestone(milestone_tracker_t* const mt,
                                     iota_milestone_t* const candidate) {
   retcode_t ret = RC_OK;
-  bool exists = false;
+  bundle_transactions_t* bundle = NULL;
+  bool exists = false, valid = false;
 
   if (candidate->index >= 0x200000) {
     return RC_OK;
   }
 
-  // Checking if milestone is already persisted/validated in database
+  // Check if milestone is already present in database i.e. validated
   TRIT_ARRAY_DECLARE(hash, NUM_TRITS_HASH);
   memcpy(hash.trits, candidate->hash, FLEX_TRIT_SIZE_243);
   if ((ret = iota_tangle_milestone_exist(mt->tangle, MILESTONE_COL_HASH, &hash,
-                                         &exists))) {
+                                         &exists)) != RC_OK) {
     goto done;
   } else if (exists) {
     goto valid;
   }
 
-  {  // TODO remove
-    iota_tangle_milestone_store(mt->tangle, candidate);
-    goto valid;
+  bundle_transactions_new(&bundle);
+  if (bundle == NULL) {
+    ret = RC_CONSENSUS_MT_OOM;
+    goto done;
+  }
+  if ((ret = bundle_validate(mt->tangle, &hash, bundle, &valid)) != RC_OK) {
+    log_warning(MILESTONE_TRACKER_LOGGER_ID, "Validating bundle failed\n");
+    goto done;
+  } else if (!valid) {
+    goto done;
+  } else {
+    trit_t signature_trits[NUM_TRITS_SIGNATURE];
+    trit_t siblings_trits[NUM_TRITS_SIGNATURE];
+    byte_t normalized_trunk[81];  // TODO size
+    trit_t normalized_trunk_trits[HASH_LENGTH];
+    trit_t sig_digest[HASH_LENGTH];
+    trit_t root[HASH_LENGTH];
+    flex_trit_t coo[FLEX_TRIT_SIZE_243];
+    iota_transaction_t tx1 = NULL;
+    iota_transaction_t tx2 = NULL;
+    Curl curl;
+
+    if ((tx1 = (iota_transaction_t)utarray_eltptr(bundle, 0)) == NULL ||
+        memcmp(tx1->hash, candidate->hash, FLEX_TRIT_SIZE_243) != 0) {
+      goto done;
+    }
+    if ((tx2 = (iota_transaction_t)utarray_eltptr(bundle, 1)) == NULL) {
+      goto done;
+    }
+
+
+    flex_trits_to_trits(signature_trits, NUM_TRITS_SIGNATURE,
+                        tx1->signature_or_message, NUM_TRITS_SIGNATURE,
+                        NUM_TRITS_SIGNATURE);
+    flex_trits_to_trits(siblings_trits, NUM_TRITS_SIGNATURE,
+                        tx2->signature_or_message, NUM_TRITS_SIGNATURE,
+                        NUM_TRITS_SIGNATURE);
+    curl.type = CURL_P_27;
+    init_curl(&curl);
+    normalize_bundle(tx1->trunk, normalized_trunk);
```

Comment:
we should rename `normalize_bundle` bundle to `normalize_hash` since we are normalizing transaction's hash too..

Reply:
Agreed